{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<KeysViewHDF5 ['_all_tag_keys', 'feature_type', 'genome', 'id', 'name']>\n",
      "_all_tag_keys\n",
      "[b'genome']\n",
      "1\n",
      "feature_type\n",
      "[b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression']\n",
      "4\n",
      "genome\n",
      "[b'Unknown' b'Unknown' b'Unknown' b'Unknown' b'Unknown']\n",
      "1\n",
      "id\n",
      "[b'ENSG00000121270' b'ENSG00000107796' b'ENSG00000163017'\n",
      " b'ENSG00000168615' b'ENSG00000123146']\n",
      "541\n",
      "name\n",
      "[b'ABCC11' b'ACTA2' b'ACTG2' b'ADAM9' b'ADGRE5']\n",
      "541\n",
      "[b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Gene Expression' b'Gene Expression'\n",
      " b'Gene Expression' b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Probe' b'Negative Control Probe'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Negative Control Codeword'\n",
      " b'Negative Control Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword' b'Blank Codeword' b'Blank Codeword'\n",
      " b'Blank Codeword' b'Blank Codeword']\n"
     ]
    }
   ],
   "source": [
    "import h5py\n",
    "import numpy as np\n",
    "\n",
    "# Open an h5 file in read mode\n",
    "file_path = 'data/hBreast/cell_feature_matrix.h5'  # Replace 'your_file.h5' with the path to your h5 file\n",
    "with h5py.File(file_path, 'r') as h5_file:\n",
    "    dataset_name = list(h5_file.keys())[0]  # Get the name of the first dataset\n",
    "    dataset = h5_file[\"/matrix/features\"]  # Access the dataset\n",
    "    print(dataset.keys())\n",
    "    for key in dataset.keys():\n",
    "        print(key)\n",
    "        print(dataset[key][:5])\n",
    "        print(len(np.unique(dataset[key][:])))\n",
    "    print(dataset[\"feature_type\"][:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     %%MatrixMarket matrix coordinate integer general\n",
      "0   %metadata_json: {\"software_version\": \"xenium u...\n",
      "1                                 541 356746 15571173\n",
      "2                                               2 1 2\n",
      "3                                               4 1 1\n",
      "4                                               6 1 9\n",
      "5                                              26 1 1\n",
      "6                                             36 1 10\n",
      "7                                              39 1 1\n",
      "8                                              59 1 1\n",
      "9                                              71 1 1\n",
      "10                                             75 1 1\n",
      "11                                            77 1 18\n",
      "12                                             86 1 7\n",
      "13                                             89 1 1\n",
      "14                                             93 1 2\n",
      "15                                             98 1 1\n",
      "16                                            104 1 4\n",
      "17                                            113 1 2\n",
      "18                                            116 1 2\n",
      "19                                            137 1 1\n",
      "20                                            153 1 1\n",
      "21                                           160 1 14\n",
      "22                                            163 1 1\n",
      "23                                            164 1 1\n",
      "24                                            165 1 1\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "\n",
    "# Define the path to the .tsv.gz file\n",
    "tsv_gz_file_path = 'data/hBreast/cell_feature_matrix/features.tsv.gz'  # Replace with the path to your .tsv.gz file\n",
    "\n",
    "# Open and read the .tsv.gz file\n",
    "with gzip.open(tsv_gz_file_path, 'rt') as file:\n",
    "    # Read the file into a pandas DataFrame\n",
    "    df = pd.read_csv(file, delimiter='\\t')\n",
    "    \n",
    "    # Display the first few rows of the DataFrame\n",
    "    print(df.head(25))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix shape: (541, 356746)\n",
      "[[0 0 0 0 0]\n",
      " [2 0 1 2 0]\n",
      " [0 0 0 1 0]\n",
      " [1 1 0 0 2]\n",
      " [0 2 0 0 0]]\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "from scipy.io import mmread\n",
    "\n",
    "# Define the path to the .mtx.gz file\n",
    "mtx_gz_file_path = 'data/hBreast/cell_feature_matrix/matrix.mtx.gz'  # Replace with the path to your .mtx.gz file\n",
    "\n",
    "# Open and read the .mtx.gz file\n",
    "with gzip.open(mtx_gz_file_path, 'rt') as file:\n",
    "    # Read the file into a sparse matrix\n",
    "    sparse_matrix = mmread(file)\n",
    "    \n",
    "    # Convert the sparse matrix to a dense format (if needed)\n",
    "    dense_matrix = sparse_matrix.toarray()\n",
    "    \n",
    "    # Display the shape of the matrix\n",
    "    print(\"Matrix shape:\", dense_matrix.shape)\n",
    "    \n",
    "    # Display the first few rows of the matrix\n",
    "    print(dense_matrix[:5, :5])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
